src/Type/Doctrine/Collection/FirstTypeSpecifyingExtension.php

```diff
@@ -0,0 +1,64 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Doctrine\Collection;
+
+use PhpParser\Node\Expr\MethodCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Analyser\SpecifiedTypes;
+use PHPStan\Analyser\TypeSpecifier;
+use PHPStan\Analyser\TypeSpecifierAwareExtension;
+use PHPStan\Analyser\TypeSpecifierContext;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Type\BooleanType;
+use PHPStan\Type\MethodTypeSpecifyingExtension;
+use PHPStan\Type\TypeCombinator;
+
+final class FirstTypeSpecifyingExtension implements MethodTypeSpecifyingExtension, TypeSpecifierAwareExtension
+{
+	private const COLLECTION_CLASS = 'Doctrine\Common\Collections\Collection';
+	private const IS_EMPTY_METHOD_NAME = 'isEmpty';
+	private const FIRST_METHOD_NAME = 'first';
+
+	/** @var TypeSpecifier */
+	private $typeSpecifier;
+
+	public function getClass(): string
+	{
+		return self::COLLECTION_CLASS;
+	}
+
+	public function isMethodSupported(
+		MethodReflection $methodReflection,
+		MethodCall $node,
+		TypeSpecifierContext $context
+	): bool {
+		return
+			(
+				$methodReflection->getDeclaringClass()->getName() === self::COLLECTION_CLASS
+				|| $methodReflection->getDeclaringClass()->isSubclassOf(self::COLLECTION_CLASS)
+			)
+			&& $methodReflection->getName() === self::IS_EMPTY_METHOD_NAME;
+	}
+
+	public function specifyTypes(
+		MethodReflection $methodReflection,
+		MethodCall $node,
+		Scope $scope,
+		TypeSpecifierContext $context
+	): SpecifiedTypes {
+		$classReflection = $methodReflection->getDeclaringClass();
+		$methodVariants = $classReflection->getNativeMethod(self::FIRST_METHOD_NAME)->getVariants();
+
+		return $this->typeSpecifier->create(
+			new MethodCall($node->var, self::FIRST_METHOD_NAME),
+			TypeCombinator::remove(ParametersAcceptorSelector::selectSingle($methodVariants)->getReturnType(), new BooleanType()),
```

Comment:
Since we're executing this now for all contexts, there should be an if-else about `$context->truthy()` and `$context->falsey()`. In the former case, `isFirst` is going to return `false`, in the second case, the `false` should be removed.

Reply:
I have added a condition, but it doesn't have any effect for some reason. The errors "Cannot call method on" still remain.